scanner/src/main/kotlin/ScanResultsStorage.kt

```diff
@@ -356,4 +361,36 @@ abstract class ScanResultsStorage {
      * Internal version of [add] that skips common sanity checks.
      */
     protected abstract fun addInternal(id: Identifier, scanResult: ScanResult): Result<Unit>
+
+    override fun read(pkg: Package, nestedProvenance: NestedProvenance): List<NestedProvenanceScanResult> =
+        when (val result = read(pkg.id)) {
+            is Success -> {
+                result.result.filter { it.provenance == nestedProvenance.root }
+                    .map { it.toNestedProvenanceScanResult(nestedProvenance) }
+            }
+
+            is Failure -> throw ScanStorageException(result.error)
+        }
+
+    override fun read(
+        pkg: Package,
+        nestedProvenance: NestedProvenance,
+        scannerCriteria: ScannerCriteria
+    ): List<NestedProvenanceScanResult> =
+        when (val result = read(pkg, scannerCriteria)) {
```

Comment:
Nit: Code duplication could be reduced; e.g. by introducing a helper function that is passed a read function as a lambda.